src/components/application_manager/test/include/application_manager/commands/commands_test.h

```diff
@@ -0,0 +1,142 @@
+#include "test/application_manager/mock_application_manager_settings.h"
+#include "application_manager/mock_application.h"
+
+namespace test {
+namespace components {
+namespace commands_test {
+
+using ::testing::ReturnRef;
+using ::testing::NiceMock;
+
+using ::utils::SharedPtr;
+using ::smart_objects::SmartObject;
+using ::application_manager::commands::MessageSharedPtr;
+using ::test::components::application_manager_test::MockApplicationManager;
+using ::test::components::application_manager_test::
+    MockApplicationManagerSettings;
+using ::application_manager::ApplicationSharedPtr;
+using ::test::components::application_manager_test::MockApplication;
+
+// Depending on the value type will be selected
+template <const bool kIf, class ThenT, class ElseT>
+struct TypeIf {
+  typedef ThenT Result;
+};
+template <class ThenT, class ElseT>
+struct TypeIf<false, ThenT, ElseT> {
+  typedef ElseT Result;
+};
+
+// If `kIsNice` is `true` then all used mock types
+// will be wrapped by a `NiceMock`
+
+enum CommandsTestMocks { kNotNice = 0, kIsNice };
+
+template <const CommandsTestMocks kIsNice = CommandsTestMocks::kNotNice>
+class CommandsTest : public ::testing::Test {
+ public:
+  typedef typename TypeIf<kIsNice,
+                          NiceMock<MockApplicationManager>,
+                          MockApplicationManager>::Result MockAppManager;
+  typedef typename TypeIf<kIsNice,
+                          NiceMock<MockApplicationManagerSettings>,
+                          MockApplicationManagerSettings>::Result
+      MockAppManagerSettings;
+
+  typedef typename TypeIf<kIsNice,
+                          NiceMock<MockApplication>,
+                          MockApplication>::Result MockApp;
+
+  typedef SharedPtr<MockApp> MockAppPtr;
+
+  virtual ~CommandsTest() {}
+
+  static MessageSharedPtr CreateMessage(
+      const smart_objects::SmartType type = smart_objects::SmartType_Null) {
+    return ::utils::MakeShared<SmartObject>(type);
+  }
+
+  static MockAppPtr CreateMockApp() {
+    return ::utils::MakeShared<MockApp>();
+  }
+
+  template <class Command>
+  SharedPtr<Command> CreateCommand(const uint32_t timeout,
+                                   MessageSharedPtr& msg) {
+    InitCommand(timeout);
+    return ::utils::MakeShared<Command>((msg ? msg : msg = CreateMessage()),
+                                        app_mngr_);
+  }
+  template <class Command>
+  SharedPtr<Command> CreateCommand(MessageSharedPtr& msg) {
+    return CreateCommand<Command>(kDefaultTimeout_, msg);
+  }
+  template <class Command>
+  SharedPtr<Command> CreateCommand(const uint32_t timeout = kDefaultTimeout_) {
+    InitCommand(timeout);
+    MessageSharedPtr msg = CreateMessage();
+    return ::utils::MakeShared<Command>(msg, app_mngr_);
+  }
+
+  enum { kDefaultTimeout_ = 100u };
+
+  MockAppManager app_mngr_;
+  MockAppManagerSettings app_mngr_set_;
+
+ protected:
+  virtual void InitCommand(const uint32_t& timeout) {
```

Comment:
@AByzhynar Method `InitCommand` get invoked in template method `CreateCommand` to prepare all mocks, before the command constructor will ask for needed data. But `default_timeout` from `ApplicationManagerSettings` was return ref to its data. So to prepare it we should guarantee that the object we want to return will still be alive when we need it. But we need it outside the `InitCommand` method. So, if `timeout` will be a value, when command constructor ask for `default_timeout`, mock will return ref on dead object.